python/cudf/cudf/_lib/join.pyx

```diff
@@ -3,220 +3,95 @@
-                rhs_view,
-                left_on_ind,
-                right_on_ind,
-                c_columns_in_common
-            ))
-    elif how == 'left':
-        with nogil:
-            c_result = move(cpp_join.left_join(
-                lhs_view,
-                rhs_view,
-                left_on_ind,
-                right_on_ind,
-                c_columns_in_common
-            ))
-    elif how == 'outer':
-        with nogil:
-            c_result = move(cpp_join.full_join(
-                lhs_view,
-                rhs_view,
-                left_on_ind,
-                right_on_ind,
-                c_columns_in_common
-            ))
-    elif how == 'leftsemi':
-        with nogil:
-            c_result = move(cpp_join.left_semi_join(
-                lhs_view,
-                rhs_view,
-                left_on_ind,
-                right_on_ind,
-                all_left_inds
-            ))
-    elif how == 'leftanti':
-        with nogil:
-            c_result = move(cpp_join.left_anti_join(
-                lhs_view,
-                rhs_view,
-                left_on_ind,
-                right_on_ind,
-                all_left_inds
-            ))
-
-    all_cols_py = columns_from_ptr(move(c_result))
-    if left_index or right_index:
-        ind_cols = OrderedDict()
-        for name, pos in zip(
-            result_index_names[::-1], result_idx_positions[::-1]
-        ):
-            ind_cols[name] = all_cols_py.pop(pos)
-        index = OrderedDict()
-        for k, v in reversed(ind_cols.items()):
-            index[k] = v
-        index = Table(index)
-    else:
-        index = None
-    data_ordered_dict = OrderedDict(zip(result_col_names, all_cols_py))
-    return Table(data=data_ordered_dict, index=index)
-
-
-def compute_result_col_names(lhs, rhs, how):
-    """
-    Determine the names of the data columns in the result of
-    a libcudf join, based on the original left and right frames
-    as well as the type of join that was performed.
-    """
-    if how in {"left", "inner", "outer", "leftsemi", "leftanti"}:
-        a = lhs._data.keys()
-        if how not in {"leftsemi", "leftanti"}:
-            return list(chain(a, (k for k in rhs._data.keys()
-                        if k not in lhs._data.keys())))
-        return list(a)
+
+cpdef semi_join(Table lhs, Table rhs, how=None):
+
+    from cudf.core.column import as_column
+
+    # left-semi and left-anti joins
+    cdef cpp_join.gather_map_type c_result
+    cdef table_view c_lhs = lhs.view()
+    cdef table_view c_rhs = rhs.view()
+
+    if how == "leftsemi":
+        c_result = move(cpp_join.left_semi_join(
+            c_lhs,
+            c_rhs
+        ))
+    elif how == "leftanti":
+        c_result = move(cpp_join.left_anti_join(
+            c_lhs,
+            c_rhs
+        ))
     else:
-        raise NotImplementedError(
-            f"{how} merge not supported yet"
-        )
+        raise ValueError(f"Invalid join type {how}")
```

Comment:
Might want to have slightly different error messages for these two functions just so there's never confusion which one is erroring.

Reply:
The traceback provides this information more reliably:

```python
In [10]: from cudf._lib.join import join, semi_join

In [11]: join(lhs, rhs, how="blah")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-11-010d26d3461f> in <module>
----> 1 join(lhs, rhs, how="blah")

~/cudf/python/cudf/cudf/_lib/join.pyx in cudf._lib.join.join()

~/cudf/python/cudf/cudf/_lib/join.pyx in cudf._lib.join.join()

ValueError: Invalid join type blah

In [12]: semi_join(lhs, rhs, how="blah")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-12-426bb13505ee> in <module>
----> 1 semi_join(lhs, rhs, how="blah")

~/cudf/python/cudf/cudf/_lib/join.pyx in cudf._lib.join.semi_join()

~/cudf/python/cudf/cudf/_lib/join.pyx in cudf._lib.join.semi_join()

ValueError: Invalid join type blah
```

Reply:
ah, yeah that's fine 👍